cdconn00/clamify-api
# [CLAM-4]

## [CLAM-4]

---

### Summary

- Configures ASP.NET Identity
- Sets up JWT Issuance/Validation
- Adds Lock Function to Swagger
- Adds secret helper provider.
- Adds missing kube files to sln file.
- Adds logging to db config.

---

### DB Changes (if any)

- [ ] - Add Identity Tables

```
CREATE TABLE IF NOT EXISTS "__EFMigrationsHistory" (
    "MigrationId" character varying(150) NOT NULL,
    "ProductVersion" character varying(32) NOT NULL,
    CONSTRAINT "PK___EFMigrationsHistory" PRIMARY KEY ("MigrationId")
);

START TRANSACTION;

DO $EF$
BEGIN
    IF NOT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname = 'Test') THEN
        CREATE SCHEMA "Test";
    END IF;
END $EF$;

DO $EF$
BEGIN
    IF NOT EXISTS(SELECT 1 FROM pg_namespace WHERE nspname = 'User') THEN
        CREATE SCHEMA "User";
    END IF;
END $EF$;

CREATE TABLE "Test"."Example" (
    "ExampleId" integer GENERATED BY DEFAULT AS IDENTITY,
    CONSTRAINT "PK_Example" PRIMARY KEY ("ExampleId")
);

CREATE TABLE "User"."Role" (
    "Id" integer GENERATED BY DEFAULT AS IDENTITY,
    "Name" character varying(256) NULL,
    "NormalizedName" character varying(256) NULL,
    "ConcurrencyStamp" text NULL,
    CONSTRAINT "PK_Role" PRIMARY KEY ("Id")
);

CREATE TABLE "User"."User" (
    "Id" integer GENERATED BY DEFAULT AS IDENTITY,
    "FirstName" text NOT NULL,
    "LastName" text NOT NULL,
    "UserName" character varying(256) NULL,
    "NormalizedUserName" character varying(256) NULL,
    "Email" character varying(256) NULL,
    "NormalizedEmail" character varying(256) NULL,
    "EmailConfirmed" boolean NOT NULL,
    "PasswordHash" text NULL,
    "SecurityStamp" text NULL,
    "ConcurrencyStamp" text NULL,
    "PhoneNumber" text NULL,
    "PhoneNumberConfirmed" boolean NOT NULL,
    "TwoFactorEnabled" boolean NOT NULL,
    "LockoutEnd" timestamp with time zone NULL,
    "LockoutEnabled" boolean NOT NULL,
    "AccessFailedCount" integer NOT NULL,
    CONSTRAINT "PK_User" PRIMARY KEY ("Id")
);

CREATE TABLE "User"."RoleClaim" (
    "Id" integer GENERATED BY DEFAULT AS IDENTITY,
    "RoleId" integer NOT NULL,
    "ClaimType" text NULL,
    "ClaimValue" text NULL,
    CONSTRAINT "PK_RoleClaim" PRIMARY KEY ("Id"),
    CONSTRAINT "FK_RoleClaim_Role_RoleId" FOREIGN KEY ("RoleId") REFERENCES "User"."Role" ("Id") ON DELETE CASCADE
);

CREATE TABLE "User"."UserClaim" (
    "Id" integer GENERATED BY DEFAULT AS IDENTITY,
    "UserId" integer NOT NULL,
    "ClaimType" text NULL,
    "ClaimValue" text NULL,
    CONSTRAINT "PK_UserClaim" PRIMARY KEY ("Id"),
    CONSTRAINT "FK_UserClaim_User_UserId" FOREIGN KEY ("UserId") REFERENCES "User"."User" ("Id") ON DELETE CASCADE
);

CREATE TABLE "User"."UserLogin" (
    "LoginProvider" text NOT NULL,
    "ProviderKey" text NOT NULL,
    "ProviderDisplayName" text NULL,
    "UserId" integer NOT NULL,
    CONSTRAINT "PK_UserLogin" PRIMARY KEY ("LoginProvider", "ProviderKey"),
    CONSTRAINT "FK_UserLogin_User_UserId" FOREIGN KEY ("UserId") REFERENCES "User"."User" ("Id") ON DELETE CASCADE
);

CREATE TABLE "User"."UserRole" (
    "UserId" integer NOT NULL,
    "RoleId" integer NOT NULL,
    CONSTRAINT "PK_UserRole" PRIMARY KEY ("UserId", "RoleId"),
    CONSTRAINT "FK_UserRole_Role_RoleId" FOREIGN KEY ("RoleId") REFERENCES "User"."Role" ("Id") ON DELETE CASCADE,
    CONSTRAINT "FK_UserRole_User_UserId" FOREIGN KEY ("UserId") REFERENCES "User"."User" ("Id") ON DELETE CASCADE
);

CREATE TABLE "User"."UserToken" (
    "UserId" integer NOT NULL,
    "LoginProvider" text NOT NULL,
    "Name" text NOT NULL,
    "Value" text NULL,
    CONSTRAINT "PK_UserToken" PRIMARY KEY ("UserId", "LoginProvider", "Name"),
    CONSTRAINT "FK_UserToken_User_UserId" FOREIGN KEY ("UserId") REFERENCES "User"."User" ("Id") ON DELETE CASCADE
);

CREATE UNIQUE INDEX "RoleNameIndex" ON "User"."Role" ("NormalizedName");

CREATE INDEX "IX_RoleClaim_RoleId" ON "User"."RoleClaim" ("RoleId");

CREATE INDEX "EmailIndex" ON "User"."User" ("NormalizedEmail");

CREATE UNIQUE INDEX "UserNameIndex" ON "User"."User" ("NormalizedUserName");

CREATE INDEX "IX_UserClaim_UserId" ON "User"."UserClaim" ("UserId");

CREATE INDEX "IX_UserLogin_UserId" ON "User"."UserLogin" ("UserId");

CREATE INDEX "IX_UserRole_RoleId" ON "User"."UserRole" ("RoleId");

INSERT INTO "__EFMigrationsHistory" ("MigrationId", "ProductVersion")
VALUES ('20221223161305_AddIdentityTables', '6.0.12');

COMMIT;
```

---

### Testing

1. Run all the unit tests.
2. Run all the integration tests.
3. Verify log table created.
4. Verify identity tables created.

---

### Post-Deployment Validation

- Verify systemlog table created in production db.
- Verify all identity tables created.

---


[CLAM-4]: https://clamify.atlassian.net/browse/CLAM-4?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
